### PR TITLE
[SYCL][PI] Add versioning for plugins

### PR DESCRIPTION
**Pull request by hdelan**
Following on from https://github.com/intel/llvm/discussions/5412

### CODE REVIEW

File: sycl/include/CL/sycl/detail/pi.h [hunk 1]
```diff
@@ -50,11 +50,36 @@
 #define _PI_H_VERSION_MAJOR 8
 #define _PI_H_VERSION_MINOR 10
 
+// Versions of plugins
+#define _PI_CUDA_PLUGIN_VERSION 1
+#define _PI_LEVEL_ZERO_PLUGIN_VERSION 1
+#define _PI_HIP_PLUGIN_VERSION 1
+#define _PI_ESIMD_PLUGIN_VERSION 1
+#define _PI_OPENCL_PLUGIN_VERSION 1
```
Comment (steffenlarsen) [inline sycl/include/CL/sycl/detail/pi.h:58]: Is there a reason to have it in pi.h rather than in the individual plugins? Apart from a handful of backend-specific enums, pi.h is backend-agnostic so if the plugin version is used by the individual plugins only it would be nicer to have them maintain them locally.

Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi.h:58]: Plugin versions should not be exposed like this. `piPluginInit` should fill this info for the user to read.

Comment (steffenlarsen) [reply sycl/include/CL/sycl/detail/pi.h:58]: @smaslov-intel is right. The plugins are separate libraries so as long as the runtime library and the plugins use the same major version of the PI API you should be able to use different versions of the plugin libraries. As such, having the plugin version in definitions in the headers may be misleading as the used plugin library may not actually be that version.

Comment (hdelan) [reply sycl/include/CL/sycl/detail/pi.h:58]: Yes that is fair enough. I don't immediately see how the user could query the Plugin version through the pi_plugin struct. Would an environment variable be more user friendly? Something like `SYCL_PI_PRINT_PLUGIN_VERSION` perhaps could print the active plugin version to screen.

Comment (steffenlarsen) [reply sycl/include/CL/sycl/detail/pi.h:58]: I think you're right that there is no current way for the user to query the version, though the information is available to the runtime through `_pi_plugin.PluginVersion`. You could make a case for `SYCL_PI_TRACE` at a certain level should report the plugin version after `piPluginInit` is called.

Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi.h:58]: What is the purpose of those prints? Who is expected to use that info and how?
SYCL_PI_TRACE=1 already prints  info about loaded plugins, so it could be extended to print basic information about the plugins loaded.

Comment (hdelan) [reply sycl/include/CL/sycl/detail/pi.h]: SYCL_PI_TRACE=1 modified to print PluginVersion.

**Comment by steffenlarsen** [comment]
CI ([6672610140](https://github.com/intel/llvm/runs/6672610140)) seems to have choked on setup. Rerunning.

File: sycl/source/detail/pi.cpp [hunk 2]
```diff
@@ -451,7 +451,8 @@ static void initializePlugins(std::vector<plugin> &Plugins) {
     if (trace(TraceLevel::PI_TRACE_BASIC))
       std::cerr << "SYCL_PI_TRACE[basic]: "
                 << "Plugin found and successfully loaded: "
-                << PluginNames[I].first << std::endl;
+                << PluginNames[I].first << " ["
+                << Plugins[I].getPiPlugin().PluginVersion << "]" << std::endl;
```
Comment (smaslov-intel) [inline sycl/source/detail/pi.cpp:455]: I suggest being a bit more verbose and say that the printed value is the version of the plugin.
Also, I think we want to "standardize" what the plugin version is, so that all plugins follow it.
So, maybe the concatenating macro should be moved to pi.h out of pi_cuda.hpp?

**Comment by cperkinsintel** [comment]
The failures in the CI are occurring when running `SYCL_PI_TRACE=-1 sycl-ls`   

I'm guessing they are legitimate failures somehow introduced by this PR.

**Comment by hdelan** [comment]
Going on holiday tomorrow - will pick this back up in 2 weeks!

**Comment by v-klochkov** [comment]
@steffenlarsen - did you try running the failed LIT test 'content.cpp' ?
It always fails on this PR and I don't see it failing on other pull-requests.

**Comment by hdelan** [comment]
> The failures in the CI are occurring when running `SYCL_PI_TRACE=-1 sycl-ls`
> 
> I'm guessing they are legitimate failures somehow introduced by this PR.

Seems to be a problem with esimd stuff. Will investigate

File: sycl/source/detail/pi.cpp [hunk 3]
```diff
@@ -452,7 +452,8 @@ static void initializePlugins(std::vector<plugin> &Plugins) {
       std::cerr << "SYCL_PI_TRACE[basic]: "
                 << "Plugin found and successfully loaded: "
                 << PluginNames[I].first << " [ PluginVersion: "
-                << Plugins[I].getPiPlugin().PluginVersion << "]" << std::endl;
+                << Plugins[PluginIdx++].getPiPlugin().PluginVersion << " ]"
```
Comment (steffenlarsen) [inline sycl/source/detail/pi.cpp:455]: Could we use `Plugins.end()` instead of `PluginIdx`? Alternatively, you could change the lines further above to something like

```c++
plugin &NewPlugin = Plugins.emplace_back(PluginInformation, PluginNames[I].second, Library);
```

and then use `NewPlugin` instead.

Comment (hdelan) [reply sycl/source/detail/pi.cpp]: Thanks. Change made

**Review by steffenlarsen — DISMISSED** [review]

File: sycl/include/CL/sycl/detail/pi.h [hunk 4]
```diff
@@ -53,6 +53,8 @@
 #define _PI_STRING_HELPER(a) #a
 #define _PI_CONCAT(a, b) _PI_STRING_HELPER(a.b)
 #define _PI_TRIPLE_CONCAT(a, b, c) _PI_STRING_HELPER(a.b.c)
+// This is the macro that plugins should all use to define their version
+#define _PI_PLUGIN_VERSION_STRING(PluginVersion) _PI_TRIPLE_CONCAT(_PI_H_VERSION_MAJOR, _PI_H_VERSION_MINOR, (PluginVersion))
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h:57]: Please use this macro in CUDA plugin to define its version

File: sycl/plugins/cuda/pi_cuda.hpp [hunk 5]
```diff
@@ -21,8 +21,7 @@
 #define _PI_CUDA_PLUGIN_VERSION 1
```
Comment (smaslov-intel) [inline sycl/plugins/cuda/pi_cuda.hpp:21]: I suggest spelling the rules how/when this version is changing.
My proposal would be to increment it by 1 with every change to the plugin.

Comment (hdelan) [reply sycl/plugins/cuda/pi_cuda.hpp:21]: Do you think this should be described in `pi_cuda.hpp` (and in every other plugin that follows this convention), or does it belong in `pi.h`?

Comment (smaslov-intel) [reply sycl/plugins/cuda/pi_cuda.hpp:21]: I think this can be suggested at pi.h level but clearly we cannot require it there.
May add a suggestion there, and a more strict guidance in each plugin.
It'd be great aid if we could somehow automate this

Comment (hdelan) [reply sycl/plugins/cuda/pi_cuda.hpp:23]: Change made. I have also added _PI_*_PLUGIN_VERSION to all plugins

Comment (smaslov-intel) [reply sycl/plugins/cuda/pi_cuda.hpp:23]: Thanks!

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp [hunk 6]
```diff
@@ -1981,16 +1981,23 @@ pi_result piTearDown(void *) {
...
 }
 
+const char SupportedVersion[] = _PI_ESIMD_PLUGIN_VERSION_STRING;
+
 pi_result piPluginInit(pi_plugin *PluginInit) {
   if (PluginInit == nullptr) {
     return PI_ERROR_INVALID_VALUE;
   }
 
+  static int PiVersionLen = strlen(PluginInit->PiVersion);
```
Comment (steffenlarsen) [inline sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp:1991]: Since we are adding more to the version string, we may want to increase the size of `PiVersion` to make room for the new version. Especially since I can imagine the plugin version increasing fairly frequently.

Comment (hdelan) [reply sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp:1991]: `PiVersion` will just hold the Major and Minor version of the PI. Whereas `PluginVersion` will hold major minor as well as the `_PI_*_PLUGIN_VERSION`. Nevertheless I have increased both to 20

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 7]
```diff
@@ -5257,11 +5257,11 @@ pi_result cuda_piextUSMGetMemAllocInfo(pi_context context, const void *ptr,
...
 pi_result cuda_piTearDown(void *) { return PI_SUCCESS; }
 
-const char SupportedVersion[] = _PI_H_VERSION_STRING;
+const char SupportedVersion[] = _PI_CUDA_PLUGIN_VERSION_STRING;
 
 pi_result piPluginInit(pi_plugin *PluginInit) {
-  int CompareVersions = strcmp(PluginInit->PiVersion, SupportedVersion);
-  if (CompareVersions < 0) {
+  static int PiVersionLen = strlen(PluginInit->PiVersion);
+  if (strncmp(PluginInit->PiVersion, SupportedVersion, PiVersionLen) < 0) {
```
[the remark this answers is not in the document]
Comment (hdelan) [reply sycl/plugins/cuda/pi_cuda.cpp:5264]: `strncmp` only compares the first `n` characters of the string, so in this case is only checking the major and minor versions (whose string uses `PiVersionLen` characters). I will change this so that it only checks the major version

File: sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp [hunk 8]
```diff
@@ -1989,7 +1989,7 @@ pi_result piPluginInit(pi_plugin *PluginInit) {
   }
 
   // Check that the major version matches in PiVersion and SupportedVersion
-  static int PiMajorVersionLen = strcspn(PluginInit->PiVersion.".");
+  static int PiMajorVersionLen = strcspn(PluginInit->PiVersion, ".");
```
Comment (smaslov-intel) [inline sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp:1992]: Other plugins need this too, and so maybe again move this to pi.h

Comment (hdelan) [reply sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp]: Change made. Using macro `_PI_VERSION_MAJOR_STRING_LEN` to define the length of Major part of string

File: sycl/include/CL/sycl/detail/pi.h [hunk 9]
```diff
@@ -66,6 +66,12 @@
 #define _PI_H_VERSION_STRING                                                   \
   _PI_CONCAT(_PI_H_VERSION_MAJOR, _PI_H_VERSION_MINOR)
 
+// _PI_VERSION_MAJOR_STRING_LEN will be used to check the major versions of
+// plugins versus the major versions of PI.
+#define _PI_STRING_SUBSTITUTE(X) _PI_STRING_HELPER(X)
+#define _PI_VERSION_MAJOR_STRING_LEN                                           \
+  sizeof(_PI_STRING_SUBSTITUTE(_PI_H_VERSION_MAJOR))
+
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h:74]: Sorry for not being clear, but I meant to implement that check here, like

#define _PI_PLUGIN_VERSION_CHECK(PiApiVersion, PiPluginVersion) \
strcmp...

or even better inline function. And call that from each plugin already, not only CUDA

Comment (hdelan) [reply sycl/include/CL/sycl/detail/pi.h]: Sure. Change made. Using macro function for consistency with other macro fctns

File: sycl/plugins/opencl/pi_opencl.hpp [hunk 10]
```diff
@@ -0,0 +1,87 @@
+//==---------- pi_opencl.hpp - OpenCL Plugin -------------------------------==//
```
Comment (smaslov-intel) [inline sycl/plugins/opencl/pi_opencl.hpp:1]: why did you need to create the pi_opencl.hpp and move some things from .cpp?

File: sycl/plugins/opencl/pi_opencl.hpp [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+#include <algorithm>
+#include <cassert>
+#include <cstring>
+#include <iostream>
+#include <limits>
+#include <map>
+#include <sstream>
+#include <string>
+#include <vector>
```
Comment (smaslov-intel) [inline sycl/plugins/opencl/pi_opencl.hpp:40]: the headers that aren't needed for this file should remain in .cpp

File: sycl/plugins/opencl/pi_opencl.hpp (hunk 10 above)
Comment (hdelan) [reply sycl/plugins/opencl/pi_opencl.hpp:1]: Just for consistency with the other plugins. Especially since the `_PI_OPENCL_PLUGIN_VERSION` I think belongs in a `hpp` file instead of a `cpp` file.

Comment (smaslov-intel) [reply sycl/plugins/opencl/pi_opencl.hpp:1]: So maybe just leave that one there and move everything else back to .cpp?

Comment (hdelan) [reply sycl/plugins/opencl/pi_opencl.hpp:1]: OK sure

Comment (smaslov-intel) [reply sycl/plugins/opencl/pi_opencl.hpp:1]: thanks

**Review by smaslov-intel — APPROVED** [review]
